app/controllers/samples_controller.rb

```diff
@@ -969,20 +969,31 @@ def results_folder
   end
 
   # GET /samples/:id/stage_results
+  # GET /samples/:id/stage_results.json
   def stage_results
-    if current_user.allowed_feature_list.include?("pipeline_viz")
-      @results = {}
-      @sample.first_pipeline_run.pipeline_run_stages.each do |stage|
+    pipeline_run = @sample.first_pipeline_run
+    feature_allowed = current_user.allowed_feature_list.include?("pipeline_viz")
+    if feature_allowed && pipeline_run
+      stage_info = {}
+      pipeline_run.pipeline_run_stages.each do |stage|
         if stage.name != "Experimental" || current_user.admin?
-          @results[stage.name] = JSON.parse stage.dag_json
+          stage_info[stage.name] = stage.dag_json.nil? ? {} : JSON.parse(stage.dag_json)
```

Comment:
nit can be more compact using `stage_info[stage.name] = JSON.PARSE(stage.dag_json || "{}")`